### PR TITLE
STYLE: upgrade ruff

### LINKED ISSUE #52448: STYLE use TypeAlias and upgrade ruff
The latest autoupdate job removes some useful type aliases definitions

As noted by @twoertwein , this is because we haven't annotated them as `TypeAlias`

So, the task here is:
1. `pre-commit autoupdate`
2. add
```
  # Useless statement
  "B018",
```
under `Additional checks that don't pass yet` in `pyproject.toml`

3. check which files were updated in https://github.com/pandas-dev/pandas/pull/52345/files . For any place where the type alias definition was replaced with `...`, you should annotate the variable with `TypeAlias`

Here's an example of the kind of change you'd be making:
```diff
diff --git a/pandas/_libs/interval.pyi b/pandas/_libs/interval.pyi
index 4c36246e04..f0412b8397 100644
--- a/pandas/_libs/interval.pyi
+++ b/pandas/_libs/interval.pyi
@@ -3,6 +3,7 @@ from typing import (
     Generic,
     TypeVar,
     overload,
+    TypeAlias,
 )
 
 import numpy as np
@@ -16,9 +17,9 @@ from pandas._typing import (
 
 VALID_CLOSED: frozenset[str]
 
-_OrderableScalarT = TypeVar("_OrderableScalarT", int, float)
-_OrderableTimesT = TypeVar("_OrderableTimesT", Timestamp, Timedelta)
-_OrderableT = TypeVar("_OrderableT", int, float, Timestamp, Timedelta)
+_OrderableScalarT: TypeAlias = TypeVar("_OrderableScalarT", int, float)
+_OrderableTimesT: TypeAlias = TypeVar("_OrderableTimesT", Timestamp, Timedelta)
+_OrderableT: TypeAlias = TypeVar("_OrderableT", int, float, Timestamp, Timedelta)
 
 class _LengthDescriptor:
     @overload
```

4. run `pre-commit run ruff --all-files`, and check that no type alias definitions are overwritten by `...`
5. run `pre-commit run --all-files`, check everything passes
6. stage, commit, open pull request

See the contributing guide for how to get started https://pandas.pydata.org/docs/dev/development/contributing.html

### PR DESCRIPTION
**Pull request by snorfyang**
- [x] closes #52448 
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).

### CODE REVIEW

**Comment by snorfyang** [comment]
There are still some errors. On the latest version of ruff, codes like 
```
match = (
            r"(No such file or directory|没有那个文件或目录|" r"File o directory non esistente)"
        )
```
will raises
```
ISC001 Implicitly concatenated string literals on one line
```
Meanwhile, if we just spilt the lines like
```
match = (
            r"(No such file or directory|没有那个文件或目录|" 
            r"File o directory non esistente)"
        )
```
It will be revised automatically by the black hook.

File: pandas/tests/io/excel/test_readers.py [hunk 1]
```diff
@@ -843,7 +843,9 @@ def test_bad_sheetname_raises(self, read_ext, sheet_name):
     def test_missing_file_raises(self, read_ext):
         bad_file = f"foo{read_ext}"
         # CI tests with other languages, translates to "No such file or directory"
-        match = r"(No such file or directory|没有那个文件或目录|File o directory non esistente)"
+        match = (
+            r"(No such file or directory|没有那个文件或目录|" r"File o directory non esistente)"
+        )
```
Comment (MarcoGorelli) [inline pandas/tests/io/excel/test_readers.py:848]: you can write this one as something like
```python
match = '|'.join([
    "No such file or directory",
    "没有那个文件或目录",
    "File o directory non esistente",
])
```
(with `black`'s reformatting)

File: pandas/core/frame.py [hunk 2]
```diff
@@ -5150,7 +5150,7 @@ def drop(
         ...                         [1, 0.8], [0.3, 0.2]])
         >>> df
                         big     small
-        lama    speed   45.0    30.0
+        llama    speed   45.0    30.0
```
Comment (MarcoGorelli) [inline pandas/core/frame.py:5153]: these aren't aligned anymore

I'd suggest either:
- anytime you change lama to llama, fixup the alignment
- or, just keep the `codespell` update out of this PR

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
thanks for working on this

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -79,12 +79,12 @@ repos:
             '--filter=-readability/casting,-runtime/int,-build/include_subdir,-readability/fn_size'
         ]
 -   repo: https://github.com/pycqa/pylint
-    rev: v2.16.2
+    rev: v3.0.0a6
```
Comment (MarcoGorelli) [inline .pre-commit-config.yaml:82]: looks like pylint have pushed a new tag, but no new release. I'd suggest reverting this change for now

File: pandas/_libs/lib.pyi [hunk 4]
```diff
@@ -30,7 +30,7 @@ from enum import Enum
 class _NoDefault(Enum):
     no_default = ...
 
-no_default: Final = _NoDefault.no_default
+no_default: Final = _NoDefault.no_default  # noqa
```
[the remark this answers is not in the document]
Comment (snorfyang) [reply pandas/_libs/lib.pyi:33]: Currently I don't have any other way to stop ruff(or black?) change its value to '''...''' even if add TypeAlias.

File: typings/numba.pyi [hunk 5]
```diff
@@ -1,3 +1,4 @@
+# ruff: noqa: PYI033
```
Comment (twoertwein) [inline typings/numba.pyi:1]: Why is this needed? Is this a bug in ruff? This file has comments, such as the typeshed-recommended `# incomplete`, but no type comments

Comment (snorfyang) [reply typings/numba.pyi:1]: The second line has a PYI033 problem.

Comment (twoertwein) [reply typings/numba.pyi:1]: That should be a bug in ruff :)

If possible (not sure whether it interferes with pyright), can you move the ruff comment after the pyright comment: `# pyright: reportIncompleteStub = false # ruff: noqa: PYI033`

Comment (snorfyang) [reply typings/numba.pyi:1]: Then Pylance raise an syntax error:
```
Pyright comment directive must be followed by "=" and a value of true, false, error, warning, information, or none
```

Comment (twoertwein) [reply typings/numba.pyi:1]: Then leave it as is.

Opened https://github.com/charliermarsh/ruff/issues/4151

File: .pre-commit-config.yaml (hunk 3 above)
Comment (MarcoGorelli) [reply .pre-commit-config.yaml:82]: ping on this

File: pandas/core/frame.py (hunk 2 above)
Comment (MarcoGorelli) [reply pandas/core/frame.py:5153]: and this

File: =1.5 [hunk 6]
```diff
@@ -0,0 +1,6 @@
+Requirement already satisfied: pandas in /home/yusharthsingh/Desktop/pandas (0+untagged.32262.g3790452.dirty)
```
Comment (MarcoGorelli) [inline =1.5:1]: remove this file

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: typings/numba.pyi (hunk 5 above)
Comment (twoertwein) [reply typings/numba.pyi:1]: a new ruff version with the fix released :)

File: pandas/core/series.py [hunk 7]
```diff
@@ -4881,7 +4881,7 @@ def drop(
         dtype: float64
 
         >>> s.drop(labels='weight', level=1)
-        lama    speed      45.0
+        llama    speed      45.0
```
Comment (MarcoGorelli) [inline pandas/core/series.py:4884]:
```suggestion
        llama   speed      45.0
```

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
small comment, rest might actually be good this time (if CI passes), thanks

**Review by MarcoGorelli — APPROVED** [review]
thanks @snorfyang !

**Review by twoertwein — APPROVED** [review]
